src/blog/first-100k-lines.vmd

```diff
@@ -0,0 +1,415 @@
+
+
+*TL;DR:* Have many links to centralized documentation.
+
+
+
+# Avoid Nondeterminism
+
+Have you ever had a bug that you could only reproduce one out of every ten tries? What an adventure!
+
+And then there's the bugs that you might only reproduce every hundredth try... *a harrowing endeavor.* You try and you try, and eventually give up and close the bug report as "Cannot Reproduce".
+
+
+This unpredictability is caused by non-determinism, in other words, *when there are some random factors that make it hard to predict whether something will happen.* For example:
+
+ * You're using a random number generator, and the bug only happens when the random number is odd.
+ * The bug only happens when thread A happens to access data before thread B.
+ * The bug only happens when network response B comes before network response A.
+ * You're doing something if C#'s string.GetHashCode is odd. [# I was shocked when string.GetHashCode broke my determinism in the 2020 7DRL Challenge.]
+ * The bug only happens when item X appears before item Y in a hash map in Golang or Rust.
+
+
+If non-determinism creeps into your program, you'll be testing your application, find a bug, and then never be able to reproduce it... so avoid it when you can!
+
+
+<slice/>
+
+
+# ...except you can't really avoid nondeterminism.
+
+"Easy for you to say, you're working on a compiler! My app needs network requests, I can't just avoid them!"
+
+
+You're right, of course! Network requests, overlapping file I/O, animations, threads, and lots of nondeterministic things might be necessary for your program. [#nondgood]
+
+
+But if we have non-determinism, how do we reproduce bugs?
+
+
+The answer: *deterministic replayability!*
+
+ * Record any sources of non-determinism into a file. For example:
+    * Record all responses you get from the network.
+    * Record the seed number you used to initialize your random number generator.
+    * Record all inputs from the user.
+ * Have a "replay mode" where you read from that file instead. For example:
+    * Instead of requesting to the network, read from that file.
+    * Instead of initializing your random number generator with the current timestamp, initialize it with the number you previously wrote to the file.
+    * Instead of reading input from the user, read what you previously wrote to the file.
+
+
+While you're developing or testing, your program should be recording to these files. When you find a bug, fire up replay mode, and enjoy the time you saved!
+
+
+In the 2020 7DRL Challenge, on the penultimate evening, I launched thousands of games automatically played by a "random AI" player. Three of them crashed, and I was able to fix the bugs because I could reproduce those crashes with deterministic replayability.
+
+
+C#, Golang, and Rust are all intentionally nondeterministic, which makes this rather difficult.
```

Comment:
I mean almost all languages, and all popular ones are nondeterminisitic. Instead I would recommend limiting nondet to the edges of the application, as an alternative to replayability. 
"Many popular languages like C#, Go, and Rust do not support replayability and I would recommend a different approach with them. Instead, ...

Reply:
Good point on almost all languages being nondeterministic, added that in. I'm not sure nondeterminism can be added to the leaves without making the rest of the program nondeterministic?